### PR TITLE
[Refactoring] Modseq strong typing

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[2decf55bf07cd830fe2a73fdad8825a1e2e4af91] [ERROR] Failures: 
[2decf55bf07cd830fe2a73fdad8825a1e2e4af91] [ERROR]   MessageUtilsTest.enrichMesageShouldEnrichUidAndModSeq:119 
[2decf55bf07cd830fe2a73fdad8825a1e2e4af91] Expecting:
[2decf55bf07cd830fe2a73fdad8825a1e2e4af91]  <ModSeq{uid=11}>
[2decf55bf07cd830fe2a73fdad8825a1e2e4af91] to be equal to:
[2decf55bf07cd830fe2a73fdad8825a1e2e4af91]  <11L>
```

[34/229]

File: mailbox/api/src/main/java/org/apache/james/mailbox/ModSeq.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+
+public class ModSeq implements Comparable<ModSeq> {
+    public static final ModSeq MAX_VALUE = of(Long.MAX_VALUE);
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/ModSeq.java]: unused?

File: mailbox/api/src/main/java/org/apache/james/mailbox/ModSeq.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        Preconditions.checkArgument(modSeq >= 0, "A modseq needs to be positive");
+        this.modSeq = modSeq;
+    }
+
+    public long asLong() {
+        return modSeq;
+    }
+
+    public ModSeq next() {
+        return new ModSeq(modSeq + 1);
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/ModSeq.java]: you don't need to check for overflow?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraModSeqProvider.java [hunk 3]
```diff
@@ -83,7 +83,7 @@ public MailboxException getUnderlying() {
         }
     }
 
-    private static final ModSeq FIRST_MODSEQ = new ModSeq(0);
+    private static final ModSeq FIRST_MODSEQ = ModSeq.MIN_VALUE;
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraModSeqProvider.java]: what about removing this constant and adding a static ModSeq.firstValue() instead?

**Comment by chibenwa** [comment]
```
[7d36efb3d304023e7defca409da5f620b920026c] [ERROR]   UpdatedFlagsTest.flagsUpdatedShouldThrowWhenMoqSeqIsNotAnInteger:64 
[7d36efb3d304023e7defca409da5f620b920026c] Expecting:
[7d36efb3d304023e7defca409da5f620b920026c]   <JsSuccess(UpdatedFlags(MessageUid{uid=123456},ModSeq{uid=35},Flags(Vector(Deleted, Seen),Vector(UserFlag(Old Flag 1))),Flags(Vector(Answered, Draft),Vector(UserFlag(New Flag 1)))),)>
[7d36efb3d304023e7defca409da5f620b920026c] to be an instance of:
[7d36efb3d304023e7defca409da5f620b920026c]   <play.api.libs.json.JsError>
[7d36efb3d304023e7defca409da5f620b920026c] but was instance of:
[7d36efb3d304023e7defca409da5f620b920026c]   <play.api.libs.json.JsSuccess>
```

[48/229]

**Comment by Arsnael** [comment]
tests are :green_apple: , deployment issue

```
[432dd5d2211959bb34c2cfdc2d5f2ea149d3a13c] [ERROR] Errors: 
[432dd5d2211959bb34c2cfdc2d5f2ea149d3a13c] [ERROR]   JamesDeploymentValidationTest>DeploymentValidation.validateDeployment:81 » InvalidServerResponse
```
test this please

**Comment by chibenwa** [comment]
```
[432dd5d2211959bb34c2cfdc2d5f2ea149d3a13c] [ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.5.1:buildTar (docker packaging) on project mock-smtp-server: com.google.cloud.tools.jib.api.InsecureRegistryException: Failed to verify the server at https://registry-1.docker.io/v2/library/openjdk/blobs/sha256:9a0b0ce99936ce4861d44ce1f193e881e5b40b5bf1847627061205b092fa7f1d because only secure connections are allowed. -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
CI restart...

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.5.1:buildTar (docker packaging) on project mock-smtp-server: com.google.cloud.tools.jib.api.InsecureRegistryException: Failed to verify the server at https://registry-1.docker.io/v2/library/openjdk/blobs/sha256:4901756f233739dc9439f352a6cf057303388521111fde1cb9d46dd5f00af484 because only secure connections are allowed. -> [Help 1]
```
test this please

**Comment by chibenwa** [comment]
Rebased to solve conflicts

**Comment by Arsnael** [comment]
Merged

File: mailbox/event/json/src/main/scala/org/apache/james/event/json/EventSerializer.scala [hunk 4]
```diff
@@ -279,11 +279,19 @@ class JsonSerialize(mailboxIdFactory: MailboxId.Factory, messageIdFactory: Messa
     case _ => JsError()
   }
   implicit val messageUidReads: Reads[MessageUid] = {
-    case JsNumber(value) => JsSuccess(MessageUid.of(value.toLong))
+    case JsNumber(value) => if (value.isValidLong) {
+        JsSuccess(MessageUid.of(value.toLong))
+      } else {
```
Comment (mbaechler) [inline mailbox/event/json/src/main/scala/org/apache/james/event/json/EventSerializer.scala]: the else is useless

Comment (chibenwa) [reply mailbox/event/json/src/main/scala/org/apache/james/event/json/EventSerializer.scala]: See https://github.com/linagora/james-project/pull/2927

**Pull request closed without merging**